spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1197,6 +1226,27 @@ private RuntimeException doInvokeRecordListener(final ConsumerRecord<K, V> recor
 			return null;
 		}
 
+		public void checkDeser(final ConsumerRecord<K, V> record, String headerName) {
+			Header header = record.headers().lastHeader(headerName);
+			if (header != null) {
+				try {
+					DeserializationException ex = (DeserializationException) new ObjectInputStream(
+							new ByteArrayInputStream(header.value())).readObject();
+					Headers headers = new RecordHeaders();
+					record.headers().forEach(h -> {
+						if (!h.key().startsWith(ErrorHandlingDeserializer2.KEY_DESERIALIZER_EXCEPTION_HEADER_PREFIX)) {
```

Comment:
Looks like you want to filter out the `ErrorHandlingDeserializer2.KEY_DESERIALIZER_EXCEPTION_HEADER_PREFIX`.
Can't we do such a logic with Java Stream?
Looks like `RecordHeaders` has a ctor like this `RecordHeaders(Iterable<Header> headers)` which can be used in the `collect(Collector<? super T, A, R> collector)` of the original stream.